MonoGame.Framework.Content.Pipeline/ContextScopeFactory.cs

```diff
@@ -0,0 +1,196 @@
+// MonoGame - Copyright (C) MonoGame Foundation, Inc
+// This file is subject to the terms and conditions defined in
+// file 'LICENSE.txt', which is part of this source code package.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Xna.Framework.Content.Pipeline;
+using MonoGame.Framework.Content.Pipeline.Builder;
+
+namespace MonoGame.Framework.Content
+{
+
+    /// <summary>
+    /// The <see cref="IContentContext"/> represents some sort of context operation's context.
+    /// Usually, this represents a <see cref="ContentImporterContext"/>
+    /// or <see cref="ContentProcessorContext"/>.
+    ///
+    /// However, because those types are part of the XNA namespace, they cannot be modified directly.
+    /// This interface is an adapter over those types.
+    /// </summary>
+    public interface IContentContext : IDisposable
+    {
+        /// <inheritdoc cref="ContentImporterContext.IntermediateDirectory"/>
+        public string IntermediateDirectory { get; }
+
+        /// <inheritdoc cref="ContentImporterContext.Logger"/>
+        public ContentBuildLogger Logger { get; }
+
+        /// <inheritdoc cref="ContentProcessorContext.SourceIdentity"/>
+        public ContentIdentity SourceIdentity { get; }
+    }
+
+    /// <summary>
+    /// <para>
+    /// The <see cref="ContextScopeFactory"/> facilitates access to a <see cref="IContentContext"/>
+    /// instance without direct access to the actual context.
+    /// </para>
+    ///
+    /// <para>
+    /// Anytime a content context operation is about to start, the operation should signal
+    /// the <see cref="BeginContext(IContentContext)"/> method.
+    /// <b> Critically </b>, the content operation must <b>dispose</b> the resulting context
+    /// when the operation has concluded.
+    /// </para>
+    ///
+    /// <para>
+    /// The most recent content operation can be retrieved with the <see cref="ActiveContext"/>
+    /// property.
+    /// </para>
+    ///
+    /// </summary>
+    public static class ContextScopeFactory
+    {
+        private static List<IContentContext> _contextStack = new List<IContentContext>(1);
```

Comment:
Yeah, just wanted to say I'm aware of an issue here. 
I wrote this unit test, and it is failing due to the contexts being wrong. If we are multi-threading the content pipeline, this is a huge issue. I'll get a fix committed asap. 
```csharp

[Test]
public async Task ContextTesting()
{
    var tasks = new List<Task>();
    long errorCount = 0;
    for (var i = 0; i < 100; i++)
    {
        var capturedIndex = i;
        var task = Task.Run(async () =>
        {
            try
            {
                // spawn a context to be used
                using var ctx = new NoOpContext(capturedIndex);
                ContextScopeFactory.BeginContext(ctx);

                // wait some "arbitrary" amount of time
                await Task.Delay(10);

                // get the context and make sure it is the same as the one we expect
                var retrievedCtx = ContextScopeFactory.ActiveContext;
                Assert.That(ctx, Is.EqualTo(retrievedCtx));
            }
            catch (Exception)
            {
                Interlocked.Increment(ref errorCount);
            }
        });
        tasks.Add(task);
    }

    await Task.WhenAll(tasks);
    Assert.That(errorCount, Is.EqualTo(0));
}

class NoOpContext : ContextScopeFactory.ContextScope
{
    public readonly int Id;
    public override string IntermediateDirectory { get; }
    public override ContentBuildLogger Logger { get; }
    public override ContentIdentity SourceIdentity { get; }

    public NoOpContext(int id)
    {
        Id = id;
    }

    public override string ToString() => "test-ctx-" + Id;
}
```

Reply:
Okay, I've added the unit test (and a second one), and modified the `ContextScopeFactory` to use `AsyncLocal<T>`. I understand this may be a controversial choice, because `AsyncLocal<T>` can be confusing to reason about. However, in this case, I think the usage is warranted. If multiple tasks for forked to do an content-import per task, then each task _may_ now have its own context.